compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/inference/FirBuilderInferenceSession.kt

```diff
@@ -216,3 +226,26 @@ class ConeComposedSubstitutor(val left: ConeSubstitutor, val right: ConeSubstitu
         return left.substituteOrNull(rightSubstitution ?: type)
     }
 }
+
+class FirStubTypeTransformer(
+    override val session: FirSession,
+    private val substitutor: ConeSubstitutor
+) : FirAbstractTreeTransformer<Void?>(phase = FirResolvePhase.IMPLICIT_TYPES_BODY_RESOLVE) {
```

Comment:
`FirAbstractTreeTransformer` is used for global stage-dependent transformers. For regular transformers you should use `FirTransformer` or `FirDefaultTransformer` (second one includes default overrides like `override transformResolvedTypeRef(...) = transformTypeRef(...)`)

Reply:
Good to learn these, thanks!